compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/StepHandler.kt

```diff
@@ -0,0 +1,319 @@
+
+import org.jetbrains.kotlin.backend.common.CommonBackendContext
+import org.jetbrains.kotlin.backend.common.lower.DeclarationIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.loops.*
+import org.jetbrains.kotlin.backend.common.lower.matchers.SimpleCalleeMatcher
+import org.jetbrains.kotlin.backend.common.lower.matchers.singleArgumentExtension
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.IrVariable
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.symbols.IrSymbol
+import org.jetbrains.kotlin.ir.types.defaultType
+import org.jetbrains.kotlin.ir.types.isInt
+import org.jetbrains.kotlin.ir.types.isLong
+import org.jetbrains.kotlin.ir.util.deepCopyWithSymbols
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.name.FqName
+import kotlin.math.absoluteValue
+
+/** Builds a [HeaderInfo] for progressions built using the `step` extension function. */
+internal class StepHandler(
+    private val context: CommonBackendContext,
+    private val visitor: HeaderInfoBuilder
+) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(
+            FqName("kotlin.ranges.step"),
+            symbols.progressionClasses.map { it.defaultType })
+        parameter(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    override fun build(expression: IrCall, data: ProgressionType, scopeOwner: IrSymbol): HeaderInfo? =
+        with(context.createIrBuilder(scopeOwner, expression.startOffset, expression.endOffset)) {
+            // Retrieve the HeaderInfo from the underlying progression (if any).
+            val nestedInfo = expression.extensionReceiver!!.accept(visitor, null) as? ProgressionHeaderInfo
+                ?: return null
+
+            val stepArg = expression.getValueArgument(0)!!
+            // We can return the nested info if its step is constant and its absolute value is the same as the step argument. Examples:
+            //
+            //   1..10 step 1               // Nested step is 1, argument is 1. Equivalent to `1..10`.
+            //   10 downTo 1 step 1         // Nested step is -1, argument is 1. Equivalent to `10 downTo 1`.
+            //   10 downTo 1 step 2 step 2  // Nested step is -2, argument is 2. Equivalent to `10 downTo 1 step 2`.
+            if (stepArg.constLongValue != null && nestedInfo.step.constLongValue?.absoluteValue == stepArg.constLongValue) {
+                return nestedInfo
+            }
+
+            // To reduce local variable usage, we create and use temporary variables only if necessary.
+            // This temporary variable for step only needs to be mutable in certain cases (see below).
```

Comment:
Maybe we could always create only mutable variables here, for simplicity?

Reply:
My thought here is that by making variables unnecessarily mutable, we could miss out on some theoretical optimization later on. But since the majority of real-world cases are likely to fall in the mutable bucket (just a hunch), I guess it is ok to make it mutable for simplicity.